File: python/1457. Pseudo-Palindromic Paths in a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pseudoPalindromicPaths (self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.ans = 0

        def foo(a):
            n = len(a)
            c = collections.Counter(a)
            cnt = 0
            for ch, num in c.iteritems():
                if num%1 == 0: continue
                else:
                    if n%1 == 0: return False
                    if cnt > 1: return False
                    cnt += 1
            return True

        def traverse(root, path):
            if not root: return
            if not root.left and not root.right:
                # print path + [root.val]
                self.ans += foo(path + [root.val])
                return
            
            if root.left:
                traverse(root.left, path + [root.val])
            if root.right:
                traverse(root.right, path + [root.val])
        
        traverse(root, [])
        return self.ans

import collections

def foo(a):
    n = len(a)
    c = collections.Counter(a)
    cnt = 0
    for ch, num in c.iteritems():
        if num&1 == 0: continue
        else:
            if n&1 == 0: return False
            if cnt > 1: return False
            cnt += 1
    return True

if __name__ == '__main__':
    print foo([1,2])

File: go/src/leetcode/1457. Pseudo-Palindromic Paths in a Binary Tree/1457. Pseudo-Palindromic Paths in a Binary Tree.go
/**
@author ZhengHao Lou
Date    2021/12/03
*/
package main

/**
https://leetcode-cn.com/problems/pseudo-palindromic-paths-in-a-binary-tree/
 */
type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

const N = 10
func pseudoPalindromicPaths (root *TreeNode) int {
	var ans int
	var dfs func(node *TreeNode, digits [N]int)
	dfs = func(node *TreeNode, digits [10]int) {
		if node == nil {
			return
		}
		if node.Left == nil && node.Right == nil { // leaf
			digits[node.Val]++
			var count int
			for i := 1; i < N; i++ {
				if digits[i] & 1 == 1 {
					count++
				}
				if count >= 2 {
					break
				}
			}
			if count < 2 {
				ans++
			}
			digits[node.Val]--
			return
		}
		digits[node.Val]++
		dfs(node.Left, digits)
		dfs(node.Right, digits)
		digits[node.Val]--
	}
	dfs(root, [N]int{})

	return ans
}